### PR TITLE
Fix file descriptor leak in the postoffice plugin

### LINKED ISSUE #1838: Rust IML agent leaks file descriptors on session destroy
IML agent is leaking file descriptors when it fails to push a message to the manager. See below the log after a couple of hours in VirtualBox.

This maybe affects overall health of the server:
```
[root@oss1 ~]# systemctl restart *iml*
Error: Too many open files
```

```
...
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.058  WARN iml_agent::reader: Got a manager read Error reqwest::Error { kind: Status(502), url: "https://adm.local/agent2/message/?server_boot_time=2020-04-29T08%3A29%3A04.000000%2B00%3A00Z&client_start_time=2020-04-29T08%3A46%3A26.431145%2B00%3A00Z" }. Will retry in 5 seconds.
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.059  INFO iml_agent::http_comms::session: Terminating all sessions
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.059  INFO iml_agent::http_comms::session: Terminating session PluginName("device")/Id("a80b9070-f075-4533-8a57-44d725909fc0")
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.059  INFO iml_agent::http_comms::session: Terminating session PluginName("action_runner")/Id("ea6f68eb-9114-4b21-b9b6-5e246beeaaf5")
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.059  INFO iml_agent::poller: session create request for ntp failed: Reqwest(reqwest::Error { kind: Status(502), url: "https://adm.local/agent2/message/" })
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.059  INFO iml_agent::http_comms::session: Terminating session PluginName("stats")/Id("7ce712d9-9b70-4ce2-9be6-3c3d85e9c3ae")
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.059  INFO iml_agent::http_comms::session: Terminating session PluginName("postoffice")/Id("4ee3c75e-55b3-4e41-8e56-cfc63e84f19d")
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.460  INFO iml_agent::poller: sending session create request for device
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.460  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("device").
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::poller: sending session create request for action_runner
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("action_runner").
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::poller: sending session create request for ostpool
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("ostpool").
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::poller: sending session create request for stats
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("stats").
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::poller: sending session create request for postoffice
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.461  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("postoffice").
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.462  INFO iml_agent::poller: session create request for action_runner failed: Reqwest(reqwest::Error { kind: Status(502), url: "https://adm.local/agent2/message/" })
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.462  INFO iml_agent::poller: session create request for ostpool failed: Reqwest(reqwest::Error { kind: Status(502), url: "https://adm.local/agent2/message/" })
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.462  INFO iml_agent::poller: session create request for stats failed: Reqwest(reqwest::Error { kind: Status(502), url: "https://adm.local/agent2/message/" })
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.462  INFO iml_agent::poller: session create request for postoffice failed: Reqwest(reqwest::Error { kind: Status(502), url: "https://adm.local/agent2/message/" })
abr 29 09:14:21 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:21.463  INFO iml_agent::poller: session create request for device failed: Reqwest(reqwest::Error { kind: Status(502), url: "https://adm.local/agent2/message/" })
abr 29 09:14:31 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:31.462  INFO iml_agent::poller: sending session create request for ntp
abr 29 09:14:31 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:31.462  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("ntp").
abr 29 09:14:31 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:31.570  INFO iml_agent::http_comms::session: Created new session PluginName("ntp")/Id("dc5265ec-9c86-40d4-8e12-9591d2f3c9c5")
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::poller: sending session create request for device
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("device").
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::poller: sending session create request for action_runner
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("action_runner").
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::poller: sending session create request for ostpool
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("ostpool").
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::poller: sending session create request for stats
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.469  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("stats").
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.470  INFO iml_agent::poller: sending session create request for postoffice
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.470  INFO iml_agent::http_comms::agent_client: Requesting new session for: PluginName("postoffice").
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.645  INFO iml_agent::http_comms::session: Created new session PluginName("ostpool")/Id("cdbf0f4e-df9b-4b61-90ee-071738ad4d34")
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.776  INFO iml_agent::action_plugins::action_plugin: Loaded the following ActionPlugins:
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: ostpool_destroy
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: create_lpurge_conf
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: get_ha_resource_list
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: crm_attribute
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: change_mcast_port
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: stop_unit
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: postoffice_add
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: configure_ntp
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: ostpool_add
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: action_check_stonith
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: remove_firewall_port
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: action_warning_stratagem
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: kernel_module_loaded
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: is_ntp_configured
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: postoffice_remove
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: stream_fidlists_stratagem
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: pcs
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: action_purge_stratagem
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: start_unit
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: get_unit_run_state
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: try_mount
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: enable_unit
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: disable_unit
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: lctl
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: create_ltuer_conf
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: package_version
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: start_scan_stratagem
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: restart_unit
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: get_kernel
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: package_installed
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: ostpool_remove
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: kernel_module_version
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: ostpool_wait
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: ostpool_create
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: add_firewall_port
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: create_lamigo_service
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::action_plugins::action_plugin: action_check_ha
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.777  INFO iml_agent::http_comms::session: Created new session PluginName("action_runner")/Id("d586134e-41ed-47ec-9b3d-1a3519a90a91")
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.863  INFO iml_agent::http_comms::session: Created new session PluginName("stats")/Id("e3e59438-92b7-4205-80c9-02f32b3ebd8c")
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: Apr 29 09:14:32.991  INFO iml_agent::http_comms::session: Created new session PluginName("device")/Id("7bdeb916-1af2-47ef-9861-171bd431bad6")
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: thread 'tokio-runtime-worker' panicked at 'Failed to initialize inotify: Os { code: 24, kind: Other, message: "Too many open files" }', iml-agent/src/daemon_plugins/postoffice.rs:39:13
abr 29 09:14:32 oss1.local iml-agent-daemon[4830]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### PR DESCRIPTION
**Pull request by ip1981**
Closes #1838.

Signed-off-by: Igor Pashev <pashev.igor@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1989)

### CODE REVIEW

**Comment by ip1981** [comment]
But still, this line is not hit https://github.com/whamcloud/integrated-manager-for-lustre/blob/f17ff3fdf16ed583a075de4978136d80dc2ddd9d/iml-agent/src/daemon_plugins/postoffice.rs#L166

File: iml-agent/src/daemon_plugins/postoffice.rs [hunk 1]
```diff
@@ -173,11 +175,14 @@ impl DaemonPlugin for PostOffice {
...
     async fn teardown(&mut self) -> Result<(), ImlAgentError> {
-        if let Some(wd) = self.wd.lock().await.0.clone() {
-            let _ = self.inotify.lock().await.rm_watch(wd);
-        }
         // drop all triggers
         self.routes.lock().await.clear();
+        let inotify = std::mem::replace(&mut *self.inotify.lock().await, None);
+        inotify
+            .unwrap()
+            .close()
```
Comment (ip1981) [inline iml-agent/src/daemon_plugins/postoffice.rs:183]: > When all file descriptors referring to an inotify instance have been closed, the underlying object and its resources are freed for reuse by the kernel; all associated watches are automatically freed. --- https://linux.die.net/man/7/inotify

File: iml-agent/src/daemon_plugins/postoffice.rs [hunk 2]
```diff
@@ -24,21 +24,17 @@ use stream_cancel::{Trigger, Tripwire};
 use tokio::{fs, net::UnixListener, sync::Mutex};
 use tokio_util::codec::{BytesCodec, FramedRead};
 
-pub struct POWD(pub Option<WatchDescriptor>);
```
Comment (jgrund) [inline iml-agent/src/daemon_plugins/postoffice.rs:27]: Is this not needed?

Comment (ip1981) [reply iml-agent/src/daemon_plugins/postoffice.rs:27]: No, not needed. I rely on the kernel not close all watches when the main handle is closed.

**Comment by ip1981** [comment]
> I'd expect `inotify` to close everything on drop, do you know why this is not the case?

I would postulate it is not dropped :)

**Comment by nlinker** [comment]
Since the issue is tricky, why not to add a description, how it is guaranteed now that the descriptors are freed?

File: iml-agent/src/daemon_plugins/postoffice.rs (hunk 1 above)
Comment (johnsonw) [reply iml-agent/src/daemon_plugins/postoffice.rs:185]: Does this solve the issue you are seeing?

Comment (ip1981) [reply iml-agent/src/daemon_plugins/postoffice.rs]: Well, yes. But not all the issues like never-ending loops.

File: iml-agent/src/daemon_plugins/postoffice.rs [hunk 3]
```diff
@@ -161,24 +163,22 @@ impl DaemonPlugin for PostOffice {
                         Err(e) => {
                             tracing::error!("Failed to open configuration {}: {}", &conf_file, e)
                         }
-                    }
+                    };
                 }
                 tracing::debug!("Ending Inotify Listen for {}", &conf_file);
```
Comment (ip1981) [inline iml-agent/src/daemon_plugins/postoffice.rs:168]:
```
jun 22 18:57:27 c1.local iml-agent-daemon[14361]: Jun 22 18:57:27.762 DEBUG iml_agent::daemon_plugins::postoffice: Ending Route for test
jun 22 18:57:27 c1.local iml-agent-daemon[14361]: Jun 22 18:57:27.762 DEBUG iml_agent::daemon_plugins::postoffice: Ending Route for foo
jun 22 18:57:27 c1.local iml-agent-daemon[14361]: Jun 22 18:57:27.762 DEBUG iml_agent::daemon_plugins::postoffice: Ending Inotify Listen for /etc/iml/postman.conf
```

**Pull request merged**